File: services/pm/api/src/main/java/com/example/demo/issue/IssueController.java
package com.example.demo.issue;

import com.example.demo.issue.dto.IssueRequest;
import com.example.demo.issue.dto.IssueResponse;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api")
public class IssueController {

    private final IssueService issueService;

    public IssueController(IssueService issueService) {
        this.issueService = issueService;
    }

    @PostMapping("/issues")
    public ResponseEntity<IssueResponse> create(@Valid @RequestBody IssueRequest request) {
        IssueResponse response = issueService.create(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/issues/{id}")
    public IssueResponse findById(@PathVariable UUID id) {
        return issueService.findById(id);
    }

    @GetMapping("/projects/{projectId}/issues")
    public List<IssueResponse> findByProject(@PathVariable UUID projectId) {
        return issueService.findByProject(projectId);
    }

    @GetMapping("/sprints/{sprintId}/issues")
    public List<IssueResponse> findBySprint(@PathVariable UUID sprintId) {
        return issueService.findBySprint(sprintId);
    }

    @PutMapping("/issues/{id}")
    public IssueResponse update(@PathVariable UUID id, @Valid @RequestBody IssueRequest request) {
        return issueService.update(id, request);
    }

    @DeleteMapping("/issues/{id}")
    public ResponseEntity<Void> delete(@PathVariable UUID id) {
        issueService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: services/pm/src/modules/issue/issue.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, HttpCode, HttpStatus, Req } from "@nestjs/common";
import type { RequestWithOrg } from "../../common/interfaces";
import { IssueService } from "./issue.service";
import { CreateIssueDto } from "./dto/create-issue.dto";
import { UpdateIssueDto } from "./dto/update-issue.dto";
import { ReorderIssueDto } from "./dto/reorder-issue.dto";
import { IssueResponseDto } from "./dto/issue-response.dto";

@Controller("api")
export class IssueController {
  constructor(private readonly issueService: IssueService) {}

  @Post("issues")
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() dto: CreateIssueDto, @Req() request: RequestWithOrg): Promise<IssueResponseDto> {
    const orgId = request.orgId;
    const userId = (request.headers["x-user-id"] as string) || "00000000-0000-0000-0000-000000000000";
    return this.issueService.create(dto, orgId, userId);
  }

  @Get("issues/:id")
  async findById(@Param("id") id: string, @Req() request: RequestWithOrg): Promise<IssueResponseDto> {
    const orgId = request.orgId;
    return this.issueService.findById(id, orgId);
  }

  @Get("projects/:projectId/issues")
  async findByProject(
    @Param("projectId") projectId: string,
    @Req() request: RequestWithOrg
  ): Promise<IssueResponseDto[]> {
    const orgId = request.orgId;
    return this.issueService.findByProject(projectId, orgId);
  }

  @Get("sprints/:sprintId/issues")
  async findBySprint(@Param("sprintId") sprintId: string, @Req() request: RequestWithOrg): Promise<IssueResponseDto[]> {
    const orgId = request.orgId;
    return this.issueService.findBySprint(sprintId, orgId);
  }

  @Put("issues/:id")
  async update(
    @Param("id") id: string,
    @Body() dto: UpdateIssueDto,
    @Req() request: RequestWithOrg
  ): Promise<IssueResponseDto> {
    const orgId = request.orgId;
    return this.issueService.update(id, dto, orgId);
  }

  @Post("projects/:projectId/issues/:issueId/reorder")
  @HttpCode(HttpStatus.NO_CONTENT)
  async reorder(
    @Param("projectId") projectId: string,
    @Param("issueId") issueId: string,
    @Body() dto: ReorderIssueDto,
    @Req() request: RequestWithOrg
  ): Promise<void> {
    const orgId = request.orgId;
    return this.issueService.reorder(projectId, issueId, dto, orgId);
  }

  @Delete("issues/:id")
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param("id") id: string, @Req() request: RequestWithOrg): Promise<void> {
    const orgId = request.orgId;
    return this.issueService.delete(id, orgId);
  }
}
